src/core/Akka/Helios.Concurrency.DedicatedThreadPool.cs

```diff
@@ -221,241 +244,492 @@ private void RequestWorker()
+                        break;
+                    }
+                }
+            }
 
-            if (work != null)
+            private void EnsureThreadRequested()
             {
-                //no local queue, write to a round-robin queue
-                //if (null == CurrentWorker)
-                //{
-                //using volatile instead of interlocked, no need to be exact, gaining 20% perf
-                unchecked
+                // There is a double counter here (_outstandingRequest and _semaphore)
+                // Unfair semaphore does not support value bigger than short.MaxValue,
+                // tring to Release more than short.MaxValue could fail miserably.
+
+                // The _outstandingRequest counter ensure that we only request a
+                // maximum of {ProcessorCount} to the semaphore.
+
+                // It's also more efficient to have two counter, _outstandingRequests is
+                // more lightweight than the semaphore.
+
+                int count = Volatile.Read(ref _outstandingRequests);
+                while (count < ProcessorCount)
                 {
-                    _index = (_index + 1);
-                    //need to wrap bitwise operations in parens to preserve order, otherwise this won't round-robin
-                    //to some actors if Settings.NumThreads is an odd number
-                    Workers[(_index & 0x7fffffff) % Settings.NumThreads].AddWork(work);
+                    int prev = Interlocked.CompareExchange(ref _outstandingRequests, count + 1, count);
+                    if (prev == count)
+                    {
+                        _semaphore.Release();
+                        break;
+                    }
+                    count = prev;
                 }
-                //}
-                //else //recursive task queue, write directly
-                //{
-                //    // send work directly to PoolWorker
-                //    // CurrentWorker.AddWork(work);
-                //}
             }
-            else
+
+            private void MarkThreadRequestSatisfied()
             {
-                throw new ArgumentNullException("work");
+                int count = Volatile.Read(ref _outstandingRequests);
+                while (count > 0)
+                {
+                    int prev = Interlocked.CompareExchange(ref _outstandingRequests, count - 1, count);
+                    if (prev == count)
+                    {
+                        break;
+                    }
+                    count = prev;
+                }
             }
-            return success;
         }
 
-        #region IDisposable members
+        #endregion
 
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
+        #region UnfairSemaphore implementation
 
-        public void Dispose(bool isDisposing)
+        [StructLayout(LayoutKind.Sequential)]
```

Comment:
Ok I had to refresh my memory about memory alignment and it has nothing to do with the LayoutKind. In x86 the object in the heap are 4 byte aligned, so depending where the object is allocated, its fields might not be 8 byte aligned.

This would have no impact if the underlying architecture was a real x86, but on a x64 architecture, the Interlocked performance is affected when the memory is not 8 byte aligned.

There are a few options:
- Leave it like that and accept the random perf in x86.
- Use a different semaphore implementation in x86
- Enforce the address of the field on a 8 byte boundary, but it may be hackish.